File: da/2025/2025-06-20_vibe/vibe.py
def ranges(numbers):
   diff = None
   for i,j in enumerate(sorted(numbers)):
       if i - j != diff:
           if diff is not None:
               yield first,last
           first = j
           diff = i -j
       last = j
   if diff is not None:
       yield first,last

def list_to_ranges(numbers):
   sorted_numbers = sorted(numbers)
   start = prev = sorted_numbers.pop(0)  # first interval
   for number in sorted_numbers:
       if number == prev + 1:  # belongs to same interval
           prev = number       # update end
           continue
       else:
           yield start, prev      # interval ended
           start = prev = number  # start new
   yield start, prev  # last interval

import random
v1 = [random.choice(range(1,999)) for _ in range(2000)]

timeit.timeit('list(list_to_ranges(v1))',globals=globals(), number=10)
timeit.timeit('list(ranges(v1))',globals=globals(), number=10)


# >>> timeit.timeit('list(list_to_ranges(v1))',globals=globals(), number=1)
# 0.00029139300022507086
# >>> timeit.timeit('list(ranges(v1))',globals=globals(), number=1)
# 0.0003507130004436476

File: da/2025/2025-06-20_vibe/vibe.jl
# from https://www.version2.dk/holdning/hjemmearbejde-vibecoding

function ranges_vector(numbers)
    a1 = sort(numbers)
    d1 = diff(a1)
    j1 = findall(d1 .> 1)
    @debug a1
    @debug j1
    Pair.([a1[1];a1[j1 .+ 1]], [a1[j1];a1[end]])
end

function ranges_asjo(numbers)
    res = Pair{Int}[]
    sorted_numbers = sort(numbers)
    start = prev = popfirst!(sorted_numbers)  # first interval
    for number in sorted_numbers
        if number <= (prev + 1)  # belongs to same interval
            prev = number       # update end
            continue
        else
            push!(res, Pair(start, prev))      # interval ended
            start = prev = number  # start new
            continue
        end
        push!(res, Pair(start, prev))  # last interval
    end
    push!(res, Pair(start, prev))  # last interval
    res
end

v1 = rand(1:999,2000)

ranges_vector(v1); ranges_asjo(v1);

@time ranges_vector(v1);
@time ranges_asjo(v1);
# julia> @time ranges_vector(v1);
#   0.000044 seconds (43 allocations: 48.070 KiB)

# julia> @time ranges_asjo(v1);
#   0.000030 seconds (128 allocations: 29.203 KiB)

